### DISCUSSION TITLE
Moving a sent message to another folder

### DISCUSSION THREAD

**Discussion by RichardD2** [microsoftgraph/msgraph-sdk-dotnet#2469] [Q&A]
I'm trying to update an application which was using IMAP and SMTP to use MS Graph instead, since the IMAP/SMTP access to Office365 is now slowing to a crawl.

The application needs to "file" sent messages into specific folders, rather than dumping them all in the "sent items" folder.

In the old version, this was simply a case of sending the message with SMTP, then connecting with IMAP, finding the target folder, and appending the message to it.

With MS Graph, things seem somewhat more complicated. The current process I'm using is:

1. Create a draft message - either a new message, a forward, or a reply;
2. Update the draft message if necessary;
3. Send the draft message;
4. Wait for the draft message to appear in the "Sent Items" folder;
5. Move the draft message to the target folder;

This *mostly* works, using the "prefer immutable IDs" request option. But occasionally, we end up with **a blank draft message** in the target folder, and no sign of the actual sent message in the "sent items" folder.

Am I doing something wrong?

Some code, for reference:

```csharp
// 1:
var draftMessage = await mailbox.Messages.PostAsync(newMessage, DefaultRequest, cancellationToken);

// 3:
await mailbox.Messages[draftMessage.Id].Send.PostAsync(DefaultRequest, cancellationToken);

// 4:
var sentItemsFolder = await mailbox.MailFolders[KnownFolders.SentItems].GetAsync(DefaultRequest, cancellationToken);

bool draftExists = false;
for (int attempt = 0; attempt < 5; attempt++)
{
    var sentItem = await mailbox.Messages[draftMessage.Id].GetAsync(DefaultRequest, cancellationToken);
    if (sentItem is not null)
    {
        if (string.Equals(sentItem.ParentFolderId, destinationFolder.Id, StringComparison.Ordinal))
        {
            // Sent item is already in the destination folder
            return;
        }
        if (string.Equals(sentItem.ParentFolderId, sentItemsFolder.Id, StringComparison.Ordinal))
        {
            draftExists = true;
            break;
        }
    }

    await Task.Delay(500, cancellationToken);
}

// 5:
if (draftExists)
{
    await mailbox.Messages[draftMessage.Id].Move.PostAsync(new() { DestinationId = destinationFolder.Id }, DefaultRequest, cancellationToken);
}
```

where `DefaultRequest` is:

```csharp
private static readonly RetryHandlerOption RetryOption = new()
{
    ShouldRetry = (delay, attempt, message) => message.StatusCode is HttpStatusCode.Conflict
        or (HttpStatusCode)423  // Locked
        or HttpStatusCode.ServiceUnavailable
        or HttpStatusCode.GatewayTimeout
        or (HttpStatusCode)429, // TooManyRequests
};

private static void DefaultRequest<T>([NotNull] RequestConfiguration<T> request) where T : class, new()
{
    request.Options.Add(RetryOption);
    request.Headers.Add("Prefer", "IdType=\"ImmutableId\"");
}
```

**Comment by andrueastman** [comment]
In these scenarios, are you able to lookup the `sentItem.ParentFolderId` and confirm what it is? Is there a pattern there?

Any chance you also post the same question at https://aka.ms/askgraph to get feedback from API owners who are more knowledgeable at the workings of the API?

**Reply by RichardD2** [reply]
I've posted the question over on the Q&A site as suggested:
[Problems moving a sent message to another folder - Microsoft Q&A](https://learn.microsoft.com/en-us/answers/questions/1662746/problems-moving-a-sent-message-to-another-folder)

As far as I can see, the `sentItem.ParentFolderId` will always be the "Sent Items" folder just before the move. I'm specifically checking `string.Equals(sentItem.ParentFolderId, sentItemsFolder.Id, StringComparison.Ordinal)` before moving it.

**Reply by RichardD2** [reply]
Seems like I've joined the 55.4K questions on that site which can't be answered. :(

**Comment by RichardD2** [comment]
Seems like there's no good solution for this issue.

So far, what we've been able to do is make two changes:

* In step 4, check the `sentItem.IsDraft` property, and avoid moving the message if it's still a draft;
* Add a separate scheduled task to run every 15 minutes, which scans the "sent items" folder for messages which should have been moved, and moves them to the appropriate folder.

Unfortunately, the only way we could get the scheduled task to work was by adding a specific string to the message subject. We can then use a regex on the subject to identify that the message should be moved, and where it should be moved to.

It's not perfect, but it seems to work (for now).
